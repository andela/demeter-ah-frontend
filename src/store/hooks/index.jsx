import { useEffect } from 'react';
import { getUser } from '../actions/signup';

export const useSetUser = ({ dispatch }) => {
  useEffect(() => {
    dispatch(getUser());
  }, []);
  return [];
};
