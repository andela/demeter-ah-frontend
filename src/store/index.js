import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import combineReducers from './reducers';
import dummyAction from './actions/Signup';

const middlewares = applyMiddleware(thunk);

const store = createStore(
  combineReducers,
  composeWithDevTools(middlewares),
);

store.dispatch(dummyAction);

export default store;
