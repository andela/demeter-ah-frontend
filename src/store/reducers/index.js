import { combineReducers } from 'redux';
import auth from './Auth';
import articleListing from './articleListing';
import resetPassword from './resetPassword';
import changePassword from './changePassword';
import articles from './articles';
import editProfile from './editProfile';
import viewProfile from './viewProfile';
import membership from './Membership';
import viewArticle from './viewArticle';
import relatedArticles from './relatedArticles';
import bookmarks from './bookmarks';
import bookmarkArticle from './bookmarkArticle';
import comments from './Comments';
import search from './search';
import notifications from './notifications';
import rate from './rateArticle';

const reducers = combineReducers({
  auth,
  resetPassword,
  changePassword,
  articles,
  editProfile,
  viewProfile,
  membership,
  articleListing,
  viewArticle,
  relatedArticles,
  bookmarks,
  bookmarkArticle,
  comments,
  search,
  notifications,
  rate
});

export default reducers;
