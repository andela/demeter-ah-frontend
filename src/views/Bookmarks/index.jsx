import React, { Fragment, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import ArticleCard from '../../components/ArticleCard';
import { viewBookmarkedArticle, viewBookmarkCleanUp } from '../../store/actions/bookmarks';
import { bookmarkArticle, bookmarkArticleCleanUp } from '../../store/actions/bookmarkArticle';
import Loader from '../../components/Loader';
import NoItem from '../../components/NoItem';

const Bookmark = (props) => {
  const {
    articles,
    bookmarkCompleted,
    bookmarkArticleAction,
  } = props;

  const [isMounted, setIsMounted] = useState(false);

  const handleBookmark = async (e) => {
    const { slug } = e.target.dataset;
    await bookmarkArticleAction(slug);
    await props.viewBookmarkedArticle();
    setIsMounted(false);
  };

  const noBookmark = (
    <div className="following">
      <ul className="list w-10/12 md:w-3/5">
        <NoItem message="You do not have any article bookmarked" />
      </ul>
    </div>
  );

  const getBookmarkedArticles = articles.length ? articles.map((article) => {
    const {
      id,
      title,
      description,
      slug,
      image,
      rating,
      readTime,
      createdAt,
      category,
      author: {
        firstName,
        lastName,
        authorImage
      }
    } = article;
    const date = new Date(createdAt);
    const createdDate = date.toLocaleDateString('default', { month: 'short', day: 'numeric', year: 'numeric' });

    const articleCategory = category ? category.name : 'Others';
    return (
      <ArticleCard
        key={id}
        title={title}
        description={description}
        image={image}
        bookmark
        slug={slug}
        onBookmark={handleBookmark}
        category={articleCategory}
        autImage={authorImage}
        authorname={`${firstName} ${lastName}`}
        readTime={readTime}
        rating={rating}
        createdAt={createdDate}
      />
    );
  }) : noBookmark;

  const fetchBookmarkedArticle = async () => {
    await props.viewBookmarkedArticle();
    setIsMounted(true);
  };

  useEffect(() => {
    fetchBookmarkedArticle();

    return () => {
      props.viewBookmarkCleanUp();
      props.bookmarkArticleCleanUp();
    };
  }, [bookmarkCompleted]);

  return (
    <Fragment>
      <div className="border-t-2 bg-purple-50 mb-6 overflow-y-auto">
        {getBookmarkedArticles}
      </div>
    </Fragment>
  );
};

const mapStateToProps = state => ({
  articles: state.bookmarks.articles,
  bookmarkCompleted: state.bookmarkArticle.isCompleted
});

const bookmarkArticleAction = slug => bookmarkArticle(slug);

export default connect(mapStateToProps, {
  viewBookmarkedArticle,
  bookmarkArticleAction,
  viewBookmarkCleanUp,
  bookmarkArticleCleanUp
})(Bookmark);
