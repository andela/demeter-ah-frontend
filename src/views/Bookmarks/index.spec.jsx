import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Bookmarks from './index';
import store from '../../store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Sign In Component', () => {
  it('Should render without errors', () => {
    const component = shallow(
      <Router>
        <Bookmarks store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render along with children componenet', () => {
    const component = mount(
      <Router>
        <Bookmarks store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  test('Should call toast isCompleted is true in signup use effect ', () => {
    const bookmarkArticleStore = mockStore({
      bookmarkArticle:
      {
        error: null,
        isLoading: false,
        isCompleted: true,
      },
      bookmarks:
      {
        error: null,
        isLoading: false,
        isCompleted: false,
        articles: [],
      }
    });
    const component = mount(
      <Router>
        <Bookmarks store={bookmarkArticleStore} />
      </Router>
    );

    expect(component).toMatchSnapshot();
  });

  test('Should run errors if error exist', () => {
    const bookmarkArticleStore = mockStore({
      bookmarkArticle:
      {
        error: null,
        isLoading: false,
        isCompleted: true,
      },
      bookmarks:
      {
        error: null,
        isLoading: false,
        isCompleted: false,
        articles: [],
      }
    });
    const component = mount(
      <Router>
        <Bookmarks store={bookmarkArticleStore} />
      </Router>
    );

    expect(component).toMatchSnapshot();
  });
});
