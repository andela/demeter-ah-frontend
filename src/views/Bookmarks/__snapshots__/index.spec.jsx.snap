// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In Component Should call toast isCompleted is true in signup use effect  1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Bookmark)
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Bookmark
        articles={Array []}
        bookmarkArticleAction={[Function]}
        bookmarkArticleCleanUp={[Function]}
        bookmarkCompleted={true}
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        viewBookmarkCleanUp={[Function]}
        viewBookmarkedArticle={[Function]}
      >
        <div
          className="border-t-2 bg-purple-50 mb-6 overflow-y-auto"
        >
          <loader>
            <div
              className="flex items-center justify-center py-4 h-72"
            >
              <img
                alt=""
                className="w-8 h-8"
                src="test-file-stub"
              />
            </div>
          </loader>
        </div>
      </Bookmark>
    </Connect(Bookmark)>
  </Router>
</BrowserRouter>
`;

exports[`Sign In Component Should render along with children componenet 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Bookmark)
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <Bookmark
        articles={Array []}
        bookmarkArticleAction={[Function]}
        bookmarkArticleCleanUp={[Function]}
        bookmarkCompleted={false}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        viewBookmarkCleanUp={[Function]}
        viewBookmarkedArticle={[Function]}
      >
        <div
          className="border-t-2 bg-purple-50 mb-6 overflow-y-auto"
        >
          <loader>
            <div
              className="flex items-center justify-center py-4 h-72"
            >
              <img
                alt=""
                className="w-8 h-8"
                src="test-file-stub"
              />
            </div>
          </loader>
        </div>
      </Bookmark>
    </Connect(Bookmark)>
  </Router>
</BrowserRouter>
`;

exports[`Sign In Component Should render without errors 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(Bookmark)
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</Router>
`;

exports[`Sign In Component Should run errors if error exist 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Bookmark)
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Bookmark
        articles={Array []}
        bookmarkArticleAction={[Function]}
        bookmarkArticleCleanUp={[Function]}
        bookmarkCompleted={true}
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        viewBookmarkCleanUp={[Function]}
        viewBookmarkedArticle={[Function]}
      >
        <div
          className="border-t-2 bg-purple-50 mb-6 overflow-y-auto"
        >
          <loader>
            <div
              className="flex items-center justify-center py-4 h-72"
            >
              <img
                alt=""
                className="w-8 h-8"
                src="test-file-stub"
              />
            </div>
          </loader>
        </div>
      </Bookmark>
    </Connect(Bookmark)>
  </Router>
</BrowserRouter>
`;
